### ISSUE TITLE
sql: change `extract` to return decimal to match Postgres 14

### ISSUE THREAD

**Issue by rafiss** [cockroachdb/cockroach#72523]
[Postgres 14 included this change](https://www.postgresql.org/docs/14/release-14.html) to `extract`:
> Change EXTRACT() to return type numeric instead of float8
> This avoids loss-of-precision issues in some usages. The old behavior can still be obtained by using the old underlying function date_part().
> Also, EXTRACT(date) now throws an error for units that are not part of the date data type.

At some point, maybe in v22.2 or v23.1, CRDB should match this.

Here are more details on how PG handled the backwards-incompat change: https://github.com/postgres/postgres/commit/a2da77cdb4661826482ebf2ddba1f953bc74afe4

> Changing the return type of an existing function is a
bit risky, so instead we do the following:  We implement a new set of
functions, which are now called "extract", in parallel to the existing
date_part functions.  They work the same way internally but use
numeric instead of float8.  The EXTRACT construct is now mapped by the
parser to these new extract functions.  That way, dumps of views
etc. from old versions (which would use date_part) continue to work
unchanged, but new uses will map to the new extract functions.
>
> Additionally, the reverse compilation of EXTRACT now reproduces the
original syntax, using the new mechanism introduced in
[40c24bf](https://github.com/postgres/postgres/commit/40c24bfef92530bd846e111c1742c2a54441c62c).

We probably shouldn't do this change until we advertise PG 14 compatibility.

Note that CRDB only added `date_part` in v22.1 with https://github.com/cockroachdb/cockroach/pull/72502

**Comment by rafiss** [comment]
We'll keep this in the backlog until we hear of a need for it.

There is risk with this project, since CRDB already has a `date_part` and an `extract` function in existing releases. If we change the return type of one of them, it's possible that any usage of the function in computed column would break unless we do more work to migrate those. Also, another chance for disruption is that any user queries that use these functions might stop working after the behavior change is made.
